### PR TITLE
utf-8 support in strings

### PR DESCRIPTION
**Pull request by liz3**
# Add Unicode support to strings

Resolves: https://github.com/dictu-lang/Dictu/issues/317

### What's Changed:
This updates the implementation of the string API to support unicode using the [
utf8.h](https://github.com/sheredom/utf8.h) library.

Theres still a few incomplete things, some fuctions do not need unicode versions(strip functions). But on the other side there are things where the library does not provide full functionality which is needed for `isUpper`/`isLower` for example which do not wrong, forcing the usage of the c std lib apis.

Should `String.len()` return byte length or character length, what if the string has invalid utf8? calculating the length upfront for large strings, always might be expensive.

And not enough tests are added.

### Type of Change:

- [ ] Bug fix
- [X] New feature
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping:

- [x] Tests have been updated to reflect the changes done within this PR (if applicable).
- [ ] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

**Comment by Jason2605** [comment]
This looks awesome and something definitely needed, thank you for this!! Sorry I've been so slow on this, been a bit hectic for me as of late!

> Should String.len() return byte length or character length

I would expect this to return the character length not the byte length. I like the `byteLen` addition! I'm more than happy with adding some extra test cases!

**Comment by liz3** [comment]
> This looks awesome and something definitely needed, thank you for this!! Sorry I've been so slow on this, been a bit hectic for me as of late!
> 
> > Should String.len() return byte length or character length
> 
> I would expect this to return the character length not the byte length. I like the `byteLen` addition! I'm more than happy with adding some extra test cases!
> 
> We will need to make sure this also works with the stdlib modules too, quite a big change!!

Hey, ive continued on this. I think the changes are needed, but i want to make clear it introduces a overhead into indexing/slicing and allocation of strings since simple indexing does not work anymore and is iterative now, the other approach is to store the strings as unicode in memory, this would imbue a massive increase in memory usage but retain indexing capabilities, im not entirely sure what the better tradeoff is here.

Ive went with approaching that most string functions will throw errors when provided with invalid utf-8, i do not know how the library would behave when given invalid utf-8. Exceptions are indexing and slicing which will fallback to the byte handling.
It might also be worth adding a function which lets the user know if the string is valid utf-8 maybe?

**Comment by liz3** [comment]
I am not entirely sure why the windows tests fail, there does not seam to be a clear test which **is** failing
Sadly at the moment i don't have access to a Window system.

**Comment by liz3** [comment]
@Jason2605 I believe this is good for review / merge now. I have fixed the issue with the windows builds, which turned out to be a error on my side but it took a good while to find. I see that theres a macos failure but i have run the tests on my mb and they all pass.

**Comment by Jason2605** [comment]
Nice one thank you! I don't have an ARM mac myself to exactly test unfortunately but will have a see if I can track someone down that does.

@briandowns be good to get your thoughts on this PR too!

**Comment by liz3** [comment]
> Nice one thank you! I don't have an ARM mac myself to exactly test unfortunately but will have a see if I can track someone down that does.
> 
> @briandowns be good to get your thoughts on this PR too!

i have a arm mac, infact i only have apple sillicon machines

**Comment by liz3** [comment]
The way runners fail also make it hard to see the error. The log does not really give you a hint or so to start debugging, but i ran all the tests on m2 air last night

**Comment by Jason2605** [comment]
> The way runners fail also make it hard to see the error. The log does not really give you a hint or so to start debugging, but i ran all the tests on m2 air last night

Oh interesting I see. Yeah the logs are no help at all really 😂 Usually when it's just a dump on code 1 it's a segfault somewhere, so will need to see if we can track it

**Comment by liz3** [comment]
> > The way runners fail also make it hard to see the error. The log does not really give you a hint or so to start debugging, but i ran all the tests on m2 air last night
> 
> Oh interesting I see. Yeah the logs are no help at all really 😂 Usually when it's just a dump on code 1 it's a segfault somewhere, so will need to see if we can track it

Aight maybe then a adresssanitizer can help to find it if its a heap corruption, Il try that

**Comment by liz3** [comment]
@Jason2605 was able to repro with a debug build and fixed it: https://github.com/dictu-lang/Dictu/pull/749/commits/f8d481843ba1c1529f572b8dc378f58dbd17b053

**Comment by liz3** [comment]
image
okay that's satisfying to be entirely honest!

**Comment by Jason2605** [comment]
Incredible stuff, this is so so so nice!!!!!

File: src/vm/datatypes/strings.c [hunk 1]
```diff
@@ -147,43 +171,44 @@ static Value splitString(DictuVM *vm, int argCount, Value *args) {
...
+        for (; tokenIndex < string->character_len && count < maxSplit; tokenIndex++) {
             count++;
-            *(tmp) = string->chars[tokenIndex];
-            *(tmp + 1) = '\0';
-            Value str = OBJ_VAL(copyString(vm, tmp, 1));
+          
+            char* next_ptr = utf8codepoint(ptr, &cp);
+
+            // *(tmp) = string->chars[tokenIndex];
+            // *(tmp + 1) = '\0';
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:191]:
```suggestion
```

File: src/vm/datatypes/strings.c [hunk 2]
```diff
@@ -239,7 +270,13 @@ static Value findString(DictuVM *vm, int argCount, Value *args) {
         runtimeError(vm, "find() takes either 1 or 2 arguments (%d given)", argCount);
         return EMPTY_VAL;
     }
-
+    {
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:273]: Do we need the extra scope here? It seems to happen throughout quite a few times

File: src/vm/datatypes/strings.c [hunk 3]
```diff
@@ -209,16 +234,22 @@ static Value splitString(DictuVM *vm, int argCount, Value *args) {
...
 
 static Value containsString(DictuVM *vm, int argCount, Value *args) {
+
     if (argCount != 1) {
         runtimeError(vm, "contains() takes 1 argument (%d given)", argCount);
         return EMPTY_VAL;
     }
-
+    {
+        ObjString *string = AS_STRING(args[0]);
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:247]: If we're checking utf8 legality would we also need to check the needle (args[1])?

File: src/vm/datatypes/strings.c [hunk 4]
```diff
@@ -280,42 +319,34 @@ static Value findLastString(DictuVM *vm, int argCount, Value *args) {
...
+        if(utf8str(ptr, ss->chars) == ptr)
+            return NUMBER_VAL(index);
+        index--;
+        if(index < 0)
+            break;
+        ptr = utf8rcodepoint(ptr, &cp);
     }
 
-    return NUMBER_VAL(found ? p-str : -1);
+    return NUMBER_VAL( -1);
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:349]:
```suggestion
    return NUMBER_VAL(-1);
```

File: src/vm/datatypes/strings.c [hunk 5]
```diff
@@ -701,31 +800,51 @@ static Value wrapString(DictuVM *vm, int argCount, Value *args) {
     }
 
     ObjString *string = AS_STRING(args[0]);
+    if(string->character_len == -1) {
+       runtimeError(vm, "String contains invalid UTF-8", argCount);
+       return EMPTY_VAL;
+   }
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:806]:
```suggestion
    }
```

File: src/vm/datatypes/strings.c [hunk 6]
```diff
@@ -701,31 +800,51 @@ static Value wrapString(DictuVM *vm, int argCount, Value *args) {
...
-    for (int cur = 0; string->chars[cur] != '\0'; cur++, count++) {
-        temp[cur] = string->chars[cur];
-
-        if (isspace(temp[cur])) {
-            last = cur;
+    char*ptr = temp;
+    for (int cur = 0; cur < string->character_len; cur++, count++) {
+        utf8_int32_t cp;
+        char* n = utf8codepoint(ptr, &cp);
+        // temp[cur] = string->chars[cur];
```
Comment (Jason2605) [inline src/vm/datatypes/strings.c:818]:
```suggestion
```

**Review by Jason2605 — COMMENTED** [review]
Few comments, most are nitpick style ones. Again thank you so much for this, I appreciate the effort that has gone into this :)

File: src/vm/datatypes/strings.c (hunk 2 above)
Comment (liz3) [reply src/vm/datatypes/strings.c]: I have refactored them to not do that anymore, it was for name shadowing but the fix was pretty easy

**Comment by liz3** [comment]
@Jason2605 i ran clang-format over the file in order to fix indentation

**Comment by Jason2605** [comment]
Noice yeah I should probably add something to the GH actions for that!

Very much appreciated, once again thank you for this!

**Pull request merged**